invenio_query_parser/utils.py

```diff
@@ -0,0 +1,52 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio-Query-Parser.
+# Copyright (C) 2014, 2015 CERN.
+#
+# Invenio-Query-Parser is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# Invenio-Query-Parser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the Free Software Foundation, Inc.,
+# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+#
+# In applying this licence, CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
+
+
+def build_valid_keywords_grammar():
+    """."""
+    try:
+        import re
+        from flask import current_app
+        from pypeg2 import attr
+
+        from .parser import (
+            KeywordQuery,
+            KeywordRule,
+            LeafRule,
+            NotKeywordValue,
+            SimpleQuery,
+            ValueQuery,
+        )
+
+        keywords_list = current_app.config.get('SEARCH_ALLOWED_KEYWORDS', [])
+
+        KeywordRule.grammar = attr('value', re.compile(
+            r"(\d\d\d\w{0,3}|%s)\b" % "|".join(keywords_list), re.I))
+
+        NotKeywordValue.grammar = attr('value', re.compile(
+            r'\b(?!\d\d\d\w{0,3}|%s:)\S+\b:' % ":|".join(keywords_list)))
+
+        SimpleQuery.grammar = attr(
+            'op', [NotKeywordValue, KeywordQuery, ValueQuery])
+    except ImportError:
```

Comment:
We've mostly been using [this pattern](https://www.quantifiedcode.com/knowledge-base/correctness/Catching%20ImportError%20to%20detect%20package%20availability/2LHXeS4F) for checking package availability these days.